### PR TITLE
[AMDGPU] SILowerSGPRSpills: do not update MRI reserve registers

### PR DESCRIPTION
**Pull request by perlfu**
VGPRs used for spilling do not require explicit reservation with MRI.  freezeReservedRegs() executed before register allocation ensures these are placed in the reserve set.

The only pass after SILowerSGPRSpills is SIPreAllocateWWMRegs which explicitly tests for interference before register allocation so should not reuse a WWM VGPR holding spill data. reserveReg prevents calculation of correct liveness for physical registers which could be used to extend SIPreAllocateWWMRegs.

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/sgpr-spill-overlap-wwm-reserve.mir [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 4
-# RUN: llc -mtriple=amdgcn-unknown-amdpal -mcpu=gfx1030 -start-before=si-lower-sgpr-spills -stop-after=prologepilog -verify-machineinstrs --stress-regalloc=2 -o - %s | FileCheck -check-prefix GCN %s
+# RUN: llc -mtriple=amdgcn-unknown-amdpal -mcpu=gfx1030 -start-before=si-lower-sgpr-spills -stop-after=prologepilog -verify-machineinstrs --stress-regalloc=5 -o - %s | FileCheck -check-prefix GCN %s
```
Comment (perlfu) [inline llvm/test/CodeGen/AMDGPU/sgpr-spill-overlap-wwm-reserve.mir:2]: To explain this test change:
--stress-regalloc=2 limits the VGPR class to 2 registers *after* reserve registers are subtracted.
Without reservation by SILowerSGPRSpills the 3 VGPRs count against the VGPR class usage, so the class size must be enlarged (to 5) to accommodate them.

Comment (perlfu) [reply llvm/test/CodeGen/AMDGPU/sgpr-spill-overlap-wwm-reserve.mir:2]: To preempt an obvious follow up question: "how does SILowerSGPRSpill exceed the register limit?"
`allocateSGPRSpillToVGPRLane` uses `findUnusedRegister` which does not consult the allocation in `RegisterClassInfo` which imposes the limit.

File: llvm/lib/Target/AMDGPU/SILowerSGPRSpills.cpp [hunk 2]
```diff
@@ -442,12 +440,5 @@ bool SILowerSGPRSpills::runOnMachineFunction(MachineFunction &MF) {
   SaveBlocks.clear();
   RestoreBlocks.clear();
 
-  // Updated the reserved registers with any physical VGPRs added for SGPR
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/SILowerSGPRSpills.cpp:445]: The phys VGPRs used for CSR SGPR spilling shouldn't be reused during SIPreAllocateWWMRegs.
I hope they will be marked reserved while getting the AllocationOrder from `RegClassInfo.runOnMachineFunction(MF)` while computing the AllocationOrder.
@arsenm any specific reason they are marked Reserved here explicitly?

File: llvm/test/CodeGen/AMDGPU/sgpr-spill-overlap-wwm-reserve.mir (hunk 1 above)
Comment (cdevadas) [reply llvm/test/CodeGen/AMDGPU/sgpr-spill-overlap-wwm-reserve.mir:2]: This option should have been better integrated into MRI to fully limit the register usage in various other passes.

**Review by arsenm — APPROVED** [review]
This shouldn't be necessary in the physical register CSR case

**Review by cdevadas — APPROVED** [review]

**Pull request merged**